bundle/edu.gemini.programexport.servlet/src/main/scala/edu/gemini/programexport/servlet/ProgramExportServlet.scala

```diff
@@ -0,0 +1,487 @@
+        case mg: ManualGroup => "guideGroup" := (te, mg)
+        case ag: AutomaticGroup.Active => "guideGroup" := (te, ag)
+        case _ => "guideGroup" := "initial"
+      }) ->:
+        ("primaryIndex" := ge.primaryIndex) ->:
+        ("userTargets" := (te, te.getUserTargets.asScala.toList)) ->:
+        ("base" := (te, asterism.t, None)) ->:
+        jEmptyObject
+    })
+
+  implicit def ProgramFieldsEncodeJson: EncodeJson[(ISPProgram, SPProgram)] =
+    EncodeJson { case (ispProg, spProg) =>
+      ("timeAccountAllocationCategories" := spProg.getTimeAcctAllocation) ->:
+        ("awardedTime" := spProg.getAwardedProgramTime.getMilliseconds) ->:
+        ("tooType" := spProg.getTooType.getDisplayValue) ->:
+        ("programMode" := spProg.getProgramMode.displayValue) ->:
+        ("isThesis" := spProg.isThesis) ->:
+        ("rolloverFlag" := spProg.getRolloverStatus) ->:
+        ("queueBand" := spProg.getQueueBand) ->:
+        ("affiliate" :=? Option(spProg.getPIAffiliate).map(_.displayValue)) ->?:
+        ("investigators" := spProg.getGsaPhase1Data.getCois.asScala.toList) ->:
+        ("piEmail" := spProg.getPIInfo.getEmail) ->:
+        ("piLastName" := spProg.getPILastName) ->:
+        ("piFirstName" := spProg.getPIFirstName) ->:
+        ("programId" := ispProg.getProgramID.toString) ->:
+        jEmptyObject
+    }
+
+  implicit def NoteEncodeJson: EncodeJson[SPNote] =
+    EncodeJson(n =>
+      ("text" := n.getNote) ->:
+        ("title" := n.getTitle) ->:
+        jEmptyObject
+    )
+
+  implicit def SiteQualityEncodeJson: EncodeJson[SPSiteQuality] =
+    EncodeJson(s =>
+      ("elevationConstraintMax" := s.getElevationConstraintMax) ->:
+        ("elevationConstraintMin" := s.getElevationConstraintMin) ->:
+        ("elevationConstraintType" := s.getElevationConstraintType.displayValue) ->:
+        ("timingWindows" := s.getTimingWindows.asScala.toList) ->:
+        ("wv" := s.getWaterVapor.displayValue) ->:
+        ("iq" := s.getImageQuality.displayValue) ->:
+        ("sb" := s.getSkyBackground.displayValue) ->:
+        ("cc" := s.getCloudCover.displayValue) ->:
+        jEmptyObject
+    )
+
+  implicit def GroupEncodeJson: EncodeJson[(ISPNode, SPGroup)] =
+    EncodeJson({ case (n, g) =>
+      ("key" := n.getNodeKey.toString) ->:
+        ("name" := g.getTitle) ->:
+        jEmptyObject
+    })
+
+  def ObservationSetupTime(ispObs: ISPObservation): Option[Duration] = {
+    val spObs = ispObs.getDataObject.asInstanceOf[SPObservation]
+    Option(SPTreeUtil.findInstrument(ispObs))
+      .flatMap(_.dataObject)
+      .map(_.asInstanceOf[SPInstObsComp]).map { i =>
+      spObs.getSetupTimeType match {
+        case SetupTime.Type.FULL => i.getSetupTime(ispObs)
+        case SetupTime.Type.REACQUISITION => i.getReacquisitionTime(ispObs)
+        case SetupTime.Type.NONE => Duration.ZERO
+      }
+    }
+  }
+
+  implicit def ObservationFieldsEncodeJson: EncodeJson[(ISPObservation, SPObservation)] = {
```

Comment:
Given an `ISPObservation` and a willingness to throw caution to the wind and cast the data object to `SPObservation` you could make an encoder for `ISPObservation`.